### PR TITLE
Add support for a response time tracer

### PR DESCRIPTION
**Pull request by karknu**
Add support for a response time tracer for chainsync requests that
should yield an immediate response.

### CODE REVIEW

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/PipelineDecision.hs [hunk 1]
```diff
@@ -179,7 +179,7 @@ pipelineDecisionLowHighMark lowMark highMark =
     goZero :: Nat Z -> WithOrigin BlockNo -> WithOrigin BlockNo
            -> (PipelineDecision Z, MkPipelineDecision)
     goZero Zero clientTipBlockNo serverTipBlockNo
-      | clientTipBlockNo `bnoPlus` 1 == serverTipBlockNo
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/PipelineDecision.hs:182]: Caught an off-by-one error here?

Comment (nfrisby) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/PipelineDecision.hs:182]: Ah, I see the separate commit message. 👍 (The PR title and description didn't lead me to expect this.)

**Comment by nfrisby** [comment]
Consensus diff LGTM 👍

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 2]
```diff
@@ -11,17 +11,24 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
...
+import Control.Tracer (Tracer, traceWith)
 import Network.TypedProtocol.Core
 import Network.TypedProtocol.Pipelined
 
 import Ouroboros.Network.Protocol.ChainSync.Type
+import Ouroboros.Network.Util.TaggedObserve
 
 
+data TraceChainsyncClientObservation = IntersectObservation ObserveIndicator
+                                     | NextObservation ObserveIndicator
```
Comment (nfrisby) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:29]: Ah, this is an observation for `RequestNext`, not just the "next observation".

That wasn't obvious to me. Would you add comments for each of the constructors, explain what events they correspond to? (I realize this is a Draft PR, and you might already be planning to do that 👍.)

**Comment by nfrisby** [comment]
I requested a couple of minor changes, but I didn't do a Review because it's just Draft. I'll keep an eye out for its promotion out of Draft.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs (hunk 2 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:29]: I will document the PR before the goes out of draft. However it is still very much a work in progress, together with the `karknu/counter` branch in cardano-node it only works correctly with a single upstream client.

**Comment by karknu** [comment]
Goes with https://github.com/input-output-hk/cardano-node/pull/2453 .

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 3]
```diff
@@ -12,18 +12,91 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
...
+import Control.Tracer (Tracer (..), traceWith)
 import Network.TypedProtocol.Core
 import Network.TypedProtocol.Pipelined
 
 import Ouroboros.Network.Protocol.ChainSync.Type
+import Ouroboros.Network.Util.TaggedObserve
+
+
+-- | TraceChainsyncClientObservation indicates which request-response exchange
+-- we're observing, either MsgFindIntersect or a non-pipelined MsgRequestNext.
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:38]:
```suggestion
-- we're observing, either 'MsgFindIntersect' or a non-pipelined 'MsgRequestNext'.
--
```

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 4]
```diff
@@ -12,18 +12,91 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
...
+                                     | NextObservation ObserveIndicator
+                                     deriving Show
+
+
+data TraceChainSyncClientTag peer = IntersectTag peer
+                                  | NextTag peer
+
+
+-- | TraceChainSyncClientReqRsp traces the start, end and duration of
+-- TraceChainSyncClientTag:ed observations.
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:49]:
```suggestion
-- | 'TraceChainSyncClientReqRsp' traces the start, end and duration of
-- 'TraceChainSyncClientTag':ed observations.
```

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 5]
```diff
@@ -12,18 +12,91 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
...
+type TraceChainSyncClientReqRsp peer =
+    TaggedObservable (TraceChainSyncClientTag peer) Time Time DiffTime
+
+
+instance (Show peer) => Show (TraceChainSyncClientTag peer) where
+    show (IntersectTag peer) = "IntersectTag " ++ show peer
+    show (NextTag peer)      = "NextTag " ++ show peer
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:56]: You can just derive this instance, a few lines above where 'TraceChainSyncClientTag` is introduced.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 6]
```diff
@@ -12,18 +12,91 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
...
+
+    return $ transform $ matchTaggedObservations
+        (\case
+            (IntersectTag _) -> atomically $ swapTMVar intersectState Nothing
+            (NextTag _)      -> atomically $ swapTMVar nextState Nothing
+        )
+        (\case
+            (IntersectTag _) ->
+                void . atomically . swapTMVar intersectState . Just
+            (NextTag _)      -> void . atomically . swapTMVar nextState . Just
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:81]: One could align the right hand sides better.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 7]
```diff
@@ -179,63 +252,68 @@ data ChainSyncInstruction header point tip
...
     -> ClientPipelinedStIdle n header point tip m a
     -> PeerSender (ChainSync header point tip)
                   AsClient StIdle n
                   (ChainSyncInstruction header point tip)
                   m a
 
-chainSyncClientPeerSender n@Zero (SendMsgRequestNext stNext stAwait) =
-
-    SenderYield
+chainSyncClientPeerSender tracer n@Zero (SendMsgRequestNext stNext stAwait) = SenderEffect $ do
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:274]: It will clutter the diff (so do that in a separate patch): it will look better if we re-indent the code and put `SenderEffect` in a new line.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 8]
```diff
@@ -179,63 +252,68 @@ data ChainSyncInstruction header point tip
...
                       ClientStNext {recvMsgRollForward} <- stAwait
-                      chainSyncClientPeerSender n
+                      chainSyncClientPeerSender tracer n
                         <$> recvMsgRollForward header tip
 
                   MsgRollBackward pRollback tip ->
                     SenderEffect $ do
                       ClientStNext {recvMsgRollBackward} <- stAwait
-                      chainSyncClientPeerSender n
+                      chainSyncClientPeerSender tracer n
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:312]: To avoid passing tracer around we could create a closure which has access to it.

File: ouroboros-network-framework/src/Ouroboros/Network/Util/TaggedObserve.hs [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    -> (s -> e -> d)
+    -> Tracer m (TaggedObservable t s e d)
+    -> Tracer m (TaggedObservable t s e d)
+matchTaggedObservations getStart putStart f tr = Tracer $ \case
+    obs@(TOStart t s) -> do
+        putStart t s
+        traceWith tr obs
+    (TOEnd t e _) -> do
+        before <- getStart t
+        traceWith tr $ TOEnd t e $ fmap (flip f e) before
```
Comment (coot) [inline ouroboros-network-framework/src/Ouroboros/Network/Util/TaggedObserve.hs:32]: What I like about this implementation is that it has the property that if we forget to put a matching `TOStart` or `TOEnd` the next computation will return `Nothing` (if I understand it correctly). Which is a nice property to have.

**Review by coot — CHANGES_REQUESTED** [review]
Very nice, just a few stylistic remarks.

**Comment by nfrisby** [comment]
Consensus side still LGTM, thanks!

File: ouroboros-network-framework/src/Ouroboros/Network/Util/TaggedObserve.hs (hunk 9 above)
Comment (karknu) [reply ouroboros-network-framework/src/Ouroboros/Network/Util/TaggedObserve.hs:32]: That behavior is the same as the untagged implementation.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs (hunk 5 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:57]: I don't so how.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs (hunk 7 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:275]: Will place each argument on its own line instead in order to avoid indenting the function one step further.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs (hunk 8 above)
Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:313]: I'm fine with passing the tracer around.

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs (hunk 5 above)
Comment (coot) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:57]:
```
data TraceChainSyncClientTag peer = IntersectTag peer
                                  | NextTag peer
  deriving Show
```

**Review by coot — APPROVED** [review]
LGTM

File: ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs [hunk 10]
```diff
@@ -12,18 +12,87 @@ module Ouroboros.Network.Protocol.ChainSync.ClientPipelined
...
+    => Tracer m (TraceChainSyncClientReqRsp peer)
+    -> peer
+    -> m (Tracer m TraceChainsyncClientObservation)
+peerReqRspTracer tracer peer = do
+    intersectState <- newTMVarIO Nothing
+    nextState <- newTMVarIO Nothing
+
+    return $ transform $ matchTaggedObservations
+        (\case
+            (IntersectTag _) -> atomically $ swapTMVar intersectState Nothing
```
Comment (coot) [inline ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:72]: Are these brackets necessary?

Comment (karknu) [reply ouroboros-network/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs:72]: No, they are not needed.

**Comment by karknu** [comment]
The idea behind this PR would be better suited to a tool like cardano-ping.
